modules/widgets/src/compass-widget.tsx

```diff
@@ -49,13 +49,16 @@ export class CompassWidget implements Widget<CompassWidgetProps> {
   element?: HTMLDivElement;
 
   constructor(props: CompassWidgetProps) {
-    this.id = props.id || 'compass';
-    this.viewId = props.viewId || null;
-    this.placement = props.placement || 'top-left';
-    props.transitionDuration = props.transitionDuration || 200;
-    props.label = props.label || 'Compass';
-    props.style = props.style || {};
-    this.props = props;
+    this.id = props.id ?? this.id;
```

Comment:
How much do we gain by copying these over to field rather than just accessing them from `this.props`? Seems like a bunch of boilerplate code for widgets that might not be necessary.

Reply:
Maybe one line of boilerplate would be saved since `id` is the only required member.

The projected widgets like `Tooltip` or `Marker` don't expose the `placement` member as a props since they're always in "fill".

I see the duplicate code. Maybe an abstract class could help?

Reply:
I don't think the case  for an abstract class is quite black and white, at least not until the amount of shared code grows further.
I agree that having all code visible and not relying on base class magic is also valuable.